### ISSUE TITLE
[Question] How Incremental Training Can be Achieved via HugeCTR' Model ?

### ISSUE THREAD

**Issue by tim5go** [NVIDIA/HugeCTR#238]
In [Nvidia's Developer Blog](https://developer.nvidia.com/blog/continuously-improving-recommender-systems-for-competitive-advantage-with-merlin-and-mlops/), it says HugeCTR' model can be updated by incremental data.
But it seems like there's no document / sample code to explain how it can be done.

The closest script I find is:
https://github.com/NVIDIA/HugeCTR/blob/master/test/pybind_test/wdl_mos_high_level.py

Am I on the right track?

**Comment by XiaoleiShi-NV** [comment]
You can find the introduction of all Python interfaces in HugeCTR from [here](https://github.com/NVIDIA/HugeCTR/blob/master/docs/python_interface.md).
Specifically, [here](https://github.com/NVIDIA/HugeCTR/blob/master/docs/python_interface.md#get_incremental_model-method) is the intro for the `get_incremental_model` API.

**Comment by tim5go** [comment]
@XiaoleiShi-NV 
I see. But would it be problematic when using MOS if the same user-item pair is updated more than once?

**Comment by XiaoleiShi-NV** [comment]
This will not happen because the `get_incremental_model` returns a **copy** of the updated embedding, which is a subset of the whole embedding used for training.

**Comment by tim5go** [comment]
@XiaoleiShi-NV 
But if the newly trained dataset is relatively small comparing to the historical dataset, are we running to "catastrophic forgetting" problem?

**Comment by XiaoleiShi-NV** [comment]
Could you explain more about the "catastrophic forgetting"?

**Comment by tim5go** [comment]
@XiaoleiShi-NV 
Yes, it roughly means overfitting on the newly added dataset, and forgetting the knowledge learned from pre-trained model.
See [Huawei's paper](https://arxiv.org/pdf/2009.02147.pdf), they use knowledge distillation to handle incremental training.

**Comment by XiaoleiShi-NV** [comment]
@tim5go 
**Regarding your question:**
From this paper, the author's solution to handle the "catastrophic forgetting" issue is:
```
We plan to implement a data reservoir to provide proper training data for incremental training. Some proportion of
data in the existing reservoir and the new data are interleaved to be the new reservoir.
```
In short, by mixing the new coming data with existing data in the preprocessing stage. I believe this solution is very specific and each user may have their own strategy. Besides, it's a stage in data preprocessing.
After this "catastrophic forgetting" issue is handled in the preprocessing, one can safely use the HCTR MOS to do the incremental training.

**More thinking:**
As shown in this paper:
```
 The implementation of data module is not finished for now and will be a part of future work to accomplish our framework.
```
So I think the experiment results shown in this paper don't involve special treatment for the "catastrophic forgetting" issue.
But from the comparison, the incremental mode outperforms the batch mode no matter whether the KD is adopted.
Based on this result, I think users should run tests and make comparisons to decide whether to handle the "catastrophic forgetting" issue additionally.

**Comment by tim5go** [comment]
@XiaoleiShi-NV 
Um.. there're 2 parts to handle "catastrophic forgetting" from this paper.

The first part is what you described - the **data reservoir**, which purely depends on the data pre-processing stage and have nothing to do with HCTR.

The second part is **knowledge distillation**, which involves changing the original loss function 
to be a linear combination of losses from student model & teacher model.

So, sadly, HCTR currently doesn't support it as it doesn't allow users to define custom loss function.

**Comment by XiaoleiShi-NV** [comment]
> HCTR currently doesn't support it as it doesn't allow users to define custom loss function.

This is a good point to follow. We will evaluate this requirement.
